### PR TITLE
Idle Clearpool strategy

### PR DESCRIPTION
**Pull request by poegva**

### CODE REVIEW

File: contracts/strategies/clearpool/IdleClearpoolStrategy.sol [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+        );
+
+        _mint(msg.sender, balanceAfter - balanceBefore);
+
+        return balanceAfter - balanceBefore;
+    }
+
+    /// @notice Change idleCDO address
+    /// @dev operation can be only done by the owner of the contract
+    function changeIdleCDO(address _idleCDO) external onlyOwner {
```
Comment (bugduino) [inline contracts/strategies/clearpool/IdleClearpoolStrategy.sol:236]: you can remove this one and keep only the `setWhitelistedCDO`

File: contracts/strategies/clearpool/IdleClearpoolStrategy.sol [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+        idleCDO = _idleCDO;
+    }
+
+    /// @notice allow to update whitelisted address
+    function setWhitelistedCDO(address _cdo) external onlyOwner {
+        require(_cdo != address(0), "IS_0");
+        idleCDO = _cdo;
+    }
+
+    function changeUniswapRouterPath(address[] memory newPath)
```
Comment (bugduino) [inline contracts/strategies/clearpool/IdleClearpoolStrategy.sol:246]: You can remove this one

File: contracts/strategies/clearpool/IdleClearpoolStrategy.sol [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+        override
+        returns (address[] memory)
+    {
+        address[] memory govTokens = new address[](1);
+        govTokens[0] = govToken;
+        return govTokens;
+    }
+
+    function getApr() external view returns (uint256) {
+        return IPoolMaster(strategyToken).getSupplyRate() * 100;
```
Comment (bugduino) [inline contracts/strategies/clearpool/IdleClearpoolStrategy.sol:149]: Is this in the format `1e18` for a `1%` apr?

**Review by bugduino — COMMENTED** [review]
Thanks for the PR, nice work for the strategy! In general it looks good, but we can simplify the logic a bit ideally to avoid having tokens in the strategy itself.

Currently `cpTokens` are held by the strategy and on deposit/redeem an ERC20 is minted/burned. For how the main `IdleCDO` contract works, it's perfectly fine in this case (where both the reward accrued and the underlying token can be exchanged in uni v2) to avoid altogether the use of `ERC20Upgradeable` and simply send cpTokens back to `msg.sender` (IdleCDO contract) on deposits and the opposite for redeems. 
In this way the strategy won't ever have any tokens at the end of each transaction and all tokens will be held by the main IdleCDO contract. The only change that this would imply, besides getting / sending cpToken to msg.sender, is in the `redeemRewards` method I guess, where you also need to first get cpTokens from msg.sender to the strategy, claim rewards and then send back cpTokens to msg.sender. We do this here for example https://github.com/Idle-Labs/idle-tranches/blob/master/contracts/strategies/idle/IdleStrategy.sol#L105-L111 (`idleTokens` work in the same way as `cpToken` and eg `cToken` from compound to my understanding)

Some minors:
- we can remove all references to uniswap v2 as that part as it entirely handled by the IdleCDO contract
- can you revert the changes of `contracts/interfaces/IConvex.sol`, `hardhat.config.js` ?

**Comment by zolotokrylin** [comment]
Clearpool internal issue: https://github.com/clearpool-finance/smart-contracts/issues/75

File: package.json [hunk 4]
```diff
@@ -57,7 +57,7 @@
     "@nomiclabs/hardhat-solhint": "^2.0.0",
     "@openzeppelin/contracts": "^4.4.2",
     "@openzeppelin/contracts-upgradeable": "^4.0.0",
-    "@openzeppelin/hardhat-upgrades": "^1.10.0",
+    "@openzeppelin/hardhat-upgrades": "^1.19.0",
```
[the remark this answers is not in the document]
Comment (poegva) [reply package.json:60]: Not really, I'll remove

**Comment by zolotokrylin** [comment]
@bugduino here is Uni v2 CPOOL/wETH pool: https://etherscan.io/address/0x99920e0ba2b8aead4d10de769b28ecc29693c011

**Pull request merged**